### PR TITLE
gh-250 added ability to add to TreeSet within ToSet

### LINKED ISSUE #250: Add the ability to change the implementation of the ToSet class depending on the set required
Currently it will support ToHashSet and ToTreeSet, with ToHashSet being the default

### PR DESCRIPTION
**Pull request by p3430233**
# Related Issue

- Resolve #250

### CODE REVIEW

**Review by GCHQDeveloper314 — CHANGES_REQUESTED** [review]

File: core/src/main/java/uk/gov/gchq/koryphe/impl/function/ToSet.java [hunk 1]
```diff
@@ -16,41 +16,93 @@
...
 
-        if (value instanceof Iterable) {
-            if (value instanceof Set) {
-                return (Set<?>) value;
+        if (null == value) {
+            if (implementation.isAssignableFrom(TreeSet.class)) {
+                return setImpl;
             }
-            return Sets.newHashSet((Iterable) value);
+            setImpl.add(null);
```
[the remark this answers is not in the document]
Comment (p3430233) [reply core/src/main/java/uk/gov/gchq/koryphe/impl/function/ToSet.java:77]: adding a null to a TreeSet will generate a NullPointerException

[the remark this answers is not in the document]
Comment (GCHQDeveloper314) [reply core/src/main/java/uk/gov/gchq/koryphe/impl/function/ToSet.java:77]: Ok, the historic behaviour of adding a `null` also seems odd, but that cannot be changed now. I would suggest reformatting the code a little to remove the nested if statement and improve clarity, something like this:
```java
if (implementation.isAssignableFrom(TreeSet.class) && (value == null)) {
  return setImpl;
} else if (implementation.isAssignableFrom(HashSet.class) && (value == null)) {
  setImpl.add(null);
```
Although, in the case where a `HashSet` was used and `value` was `null` the final `if` will add a it anyway. So we don't need to check for `HashSet`. The first `if` can be replaced with just:
```java
if (implementation.isAssignableFrom(TreeSet.class) && (null == value)) {
   return setImpl;
```

File: core/src/main/java/uk/gov/gchq/koryphe/impl/function/ToSet.java [hunk 2]
```diff
@@ -16,41 +16,93 @@
...
+        final Set<Object> setImpl;
 
-        if (value instanceof Object[]) {
-            return Sets.newHashSet((Object[]) value);
+        if (implementation.isAssignableFrom(HashSet.class)) {
+            setImpl = new HashSet<>();
+        } else if (implementation.isAssignableFrom(TreeSet.class)) {
+            setImpl = new TreeSet<>();
+        } else {
+            throw new IllegalArgumentException("Unrecognised Set implementation");
```
[the remark this answers is not in the document]
Comment (t92549) [reply core/src/main/java/uk/gov/gchq/koryphe/impl/function/ToSet.java:70]: Test added

File: core/src/main/java/uk/gov/gchq/koryphe/impl/function/ToSet.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (t92549) [reply core/src/main/java/uk/gov/gchq/koryphe/impl/function/ToSet.java]: Your second suggestion was used

File: core/src/main/java/uk/gov/gchq/koryphe/impl/function/ToSet.java [hunk 3]
```diff
@@ -16,41 +16,93 @@
...
+
+    @JsonInclude(JsonInclude.Include.NON_DEFAULT)
+    private Class implementation;
+
+    public ToSet()  {
+        setImplementation(DEFAULT_IMPLEMENTATION);
+    }
+
+    public ToSet(final String implementationString) throws ClassNotFoundException {
+        setImplementation(Class.forName(implementationString));
```
[the remark this answers is not in the document]
Comment (t92549) [reply core/src/main/java/uk/gov/gchq/koryphe/impl/function/ToSet.java]: As far as I understand, both of your examples will use `DEFAULT_IMPLEMENTATION` and I am not sure this is an issue?

[the remark this answers is not in the document]
Comment (GCHQDeveloper314) [reply core/src/main/java/uk/gov/gchq/koryphe/impl/function/ToSet.java]: I've just tested it (with the code as it was when I added the comment) and the first one throws an exception, the second one doesn't. This has been fixed as the latest commit by @p3430233 removes the wrapped `Class.forName()` and so there is no longer an exception thrown and the behaviour is consistent.

**Review by GCHQDeveloper314 — APPROVED** [review]

File: core/src/test/java/uk/gov/gchq/koryphe/impl/function/ToSetTest.java [hunk 4]
```diff
@@ -97,6 +156,77 @@ public void shouldReturnAGivenSet() {
...
+    public void shouldThrowExceptionWhenUnrecognisedSetImplementationUsed() {
+        // Given
+        final ToSet function = new ToSet(LinkedHashSet.class);
+
+        // When / Then
+        IllegalArgumentException e = assertThrows(IllegalArgumentException.class, () -> function.apply(null));
+        assertEquals("Unrecognised Set implementation", e.getMessage());
+    }
+
     @Override
```
Comment (t92549) [inline core/src/test/java/uk/gov/gchq/koryphe/impl/function/ToSetTest.java:230]: You could add the implementation field to `shouldJsonSerialiseAndDeserialise` test to make sure it works as expected

**Review by t92549 — APPROVED** [review]

**Pull request merged**